SetupDataPkg/ConfProfileMgrDxe/ConfProfileMgrDxe.c

```diff
@@ -0,0 +1,248 @@
+      }
+    }
+
+    FreePool (VarList);
+  }
+
+  return;
+}
+
+/**
+  This driver will query the active configuration profile and validate the flash contents match. If not, the
+  profile contents will be written to flash and the system reset.
+
+  @param  ImageHandle  Handle to this image.
+  @param  SystemTable  Pointer to the system table
+
+  @retval  EFI_SUCCESS   The driver validated the profile against flash contents successfully.
+  @retval  !EFI_SUCCESS  The driver failed to validate the profile against flash contents successfully.
+**/
+EFI_STATUS
+EFIAPI
+ConfProfileMgrDxeEntry (
+  IN  EFI_HANDLE        ImageHandle,
+  IN  EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_GUID    *ActiveProfileGuid = NULL;
+  EFI_STATUS  Status;
+  UINTN       Size = sizeof (EFI_GUID);
+  UINT32      Attributes;
+  EFI_GUID    CachedProfile;
+  BOOLEAN     NeedToFreeMem = TRUE;
+
+  // RetrieveActiveProfileGuid will validate the profile guid is valid before returning
+  Status = RetrieveActiveProfileGuid (&ActiveProfileGuid);
+
+  // Check if the status failed, but don't error out, try to use cached profile
+  if (EFI_ERROR (Status) || (ActiveProfileGuid == NULL)) {
+    NeedToFreeMem = FALSE;
+    DEBUG ((DEBUG_ERROR, "%a RetrieveActiveProfileGuid failed (%r)! Attempting to set cached profile\n", __FUNCTION__, Status));
+
+    Status = gRT->GetVariable (
+                    CACHED_CONF_PROFILE_VARIABLE_NAME,
+                    &gConfProfileMgrVariableGuid,
+                    &Attributes,
+                    &Size,
+                    &CachedProfile
+                    );
+
+    if (EFI_ERROR (Status) || (Size != sizeof (EFI_GUID))) {
+      DEBUG ((DEBUG_ERROR, "%a failed to read cached profile, defaulting to default profile (%r)!\n", __FUNCTION__, Status));
+      ActiveProfileGuid = (EFI_GUID *)&gSetupDataPkgGenericProfileGuid;
+    } else {
+      ActiveProfileGuid = &CachedProfile;
+    }
+  }
+
+  // Set the PCD for the profile parser lib to read during profile validation and for the ConfApp
+  Status = PcdSetPtrS (PcdSetupConfigActiveProfileFile, &Size, ActiveProfileGuid);
+
+  // If we fail to set the PCD, the system will still be in the correct profile, but the ConfApp will report that
+  // the generic profile is active, so we still want to validate the active profile
+  if (EFI_ERROR (Status) || (Size != sizeof (EFI_GUID))) {
+    DEBUG ((DEBUG_ERROR, "%a failed to set ActiveProfile PCD (%r)!\n", __FUNCTION__, Status));
+    ASSERT (FALSE);
+  }
+
+  // We have the active profile guid, we need to load it and validate the contents against flash.
+  // ValidateProfile does not return a status, in case of failure, it writes the chosen profile to flash
+  // and resets the system. Only validate the profile if we are in CUSTOMER_MODE (not manufacturing mode),
+  // as in debug and bringup scenarios the profile may be expected to not match
+  if (!IsSystemInManufacturingMode ()) {
+    ValidateActiveProfile ();

```

Comment:
I think there is the possibility that the system could be configured in such a way that there is no active profile, or that the active profile does not contain a corresponding set of settings.  

i.e. some entity comes along as sets a cached config profile variable name that is an invalid guid.

I think the check needs to occur that there is no active profile and that some sort of recovery would be good.

Reply:
The logic above this is intended to handle that flow (getting the active profile fails or gives a garbage value, in which case fail to the cached profile, otherwise fail to the generic profile). However, you are correct that the check is in the null lib if we are getting a legitimate profile value. At the very least, that check needs to be added for the cached profile guid, too. I agree about shifting the check out of the lib and into the profile driver.

The build will fail if SetupDataPkg is used and there is not a generic profile provisioned (and the profiles are architected such that the generic profile contains the super set of config knobs, all other profiles will be the same number of knobs, but may have different values.